File: sdks/python/src/opik/evaluation/metrics/llm_judges/hallucination/parser.py
import logging
from opik import exceptions, logging_messages
from opik.evaluation.metrics import score_result
from opik.evaluation.metrics.llm_judges import parsing_helpers

LOGGER = logging.getLogger(__name__)


def parse_model_output(content: str, name: str) -> score_result.ScoreResult:
    try:
        dict_content = parsing_helpers.extract_json_content_or_raise(content)
        score = float(dict_content["score"])

        if not (0.0 <= score <= 1.0):
            raise exceptions.MetricComputationError(
                f"Hallucination score must be between 0.0 and 1.0, got {score}"
            )

        return score_result.ScoreResult(
            name=name,
            value=score,
            reason=str(dict_content["reason"]),
        )
    except Exception as e:
        LOGGER.error(f"Failed to parse model output: {e}", exc_info=True)
        raise exceptions.MetricComputationError(
            logging_messages.HALLUCINATION_DETECTION_FAILED
        )

File: sdks/typescript/src/opik/evaluation/metrics/llmJudges/moderation/parser.ts
import { EvaluationScoreResult } from "@/evaluation/types";
import { MetricComputationError } from "../../errors";
import { extractJsonContentOrRaise } from "../parsingHelpers";
import { logger } from "@/utils/logger";

/**
 * Parses the LLM model output for the Moderation metric.
 *
 * @param content - The raw string output from the model
 * @param name - The name of the metric (for the result object)
 * @returns A score result with value and reason
 * @throws {MetricComputationError} If parsing fails or score is invalid
 */
export function parseModelOutput(
  content: string,
  name: string
): EvaluationScoreResult {
  try {
    const dictContent = extractJsonContentOrRaise(content) as Record<
      string,
      unknown
    >;

    const score = Number(dictContent["score"]);

    if (isNaN(score) || score < 0.0 || score > 1.0) {
      throw new Error(
        `Moderation score must be between 0.0 and 1.0, got ${score}`
      );
    }

    return {
      name,
      value: score,
      reason: String(dictContent["reason"] || ""),
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    logger.error(`Failed to parse model output: ${errorMessage}`);

    throw new MetricComputationError(
      "Failed to calculate moderation score. The model output could not be parsed.",
      error instanceof Error ? error : undefined
    );
  }
}
